CIP-0164/README.md

```diff
@@ -0,0 +1,2808 @@
+   certificates and not unduly increase their size.
+
+2. **Key rotation:** The cryptographic keys used to sign Leios votes and
+   certificates _do not_ need to be rotated periodically because the constraints
+   on Leios voting rounds and the key rotation already present in Praos secure
+   the protocol against attacks such as replay and key compromise.
+
+3. **Deterministic signatures:** While deterministic signatures can provide
+   additional protection against attacks that exploit weak randomness in
+   signature generation, they are not strictly required for protocol security.
+   The main requirement for deterministic randomness is in the lottery
+   mechanism, which is satisfied by the use of VRFs.
+
+4. **Local lottery:** Selection of the voting committee should not be so
+   deterministic and public as to open attack avenues such as denial-of-service
+   or subversion.
+
+5. **Liveness:** Adversaries with significant stake (e.g., more than 35%) should
+   not be able to thwart an honest majority from reaching a quorum of votes for
+   an EB.
+
+6. **Soundness:** Adversaries with near majority stake (e.g., 49%) should not be
+   able to form an adversarial quorum that certifies the EB of their choice.
+
+7. **Small votes:** Because vote traffic is large and frequent in Leios, the
+   votes themselves should be small. Note that the large size of Praos KES
+   signatures precludes their use for signing Leios votes.
+
+8. **Small certificates:** Because Leios certificates are frequent and must fit
+   inside Praos blocks, they should be small enough so there is plenty of room
+   for other transactions in the Praos blocks. Note once again that certificates
+   based on Praos KES signatures are too large for consideration in Leios.
+
+9. **Fast cryptography:** The computational burden of creating and verifying
+   voting eligibility, the votes themselves, and the resulting certificate must
+   be small enough to fit within the CPU budget for Leios stages.
+
+The aggregate signature scheme implementation specified in this document (using
+BLS as an example) satisfies all these requirements, as evidenced by the
+performance characteristics and certificate sizes documented in the
+[Votes and Certificates](#votes-and-certificates) section.
+
+<h3 id="appendix-b-cddl">Appendix B: Wire Format Specifications (CDDL)</h2>
+
+This appendix contains the complete CDDL specifications for all Leios protocol
+messages and data structures. These definitions specify the exact wire format
+for network communication.
+
+<a id="ranking-block-cddl" href="#ranking-block-cddl">**Ranking Block**</a>
+
+```diff
+ ranking_block =
+   [ header                   : block_header
+   , transaction_bodies       : [* transaction_body]
+   , transaction_witness_sets : [* transaction_witness_set]
+   , auxiliary_data_set       : {* transaction_index => auxiliary_data}
++  , ? eb_certificate         : leios_certificate
```

Comment:
I'm giving it a try at implementing Leios primitives at Pallas.
I see that `invalid_transactions` is missing here; did you think of keeping that field in the last position, or before `eb_certificate`?